core/node/node_framework/src/implementations/layers/eth_sender.rs

```diff
@@ -103,6 +110,14 @@ impl WiringLayer for EthSenderLayer {
             eth_tx_manager_actor,
         }));
 
+        // Insert circuit breaker.
+        let CircuitBreakerCheckerResource(circuit_breaker_checker) = context.get_resource().await?;
```

Comment:
This won't work if we didn't add a circuit breaker layer. Circuit breaker component is optional -- it's a defensive measure, so we can perfectly run without it, if we ever want to.

For this to work, we need to split the circuit breaker into two parts:

1. Create structure `CircuitBreakers` which will hold the list of circuit breakers we have. It will have methods related to adding circuit breakers (like `insert_breaker_if_not_exists`).
2. Implement `Default` for `CircuitBreakers`. Change `CircuitBreakerCheckerResource` to `CircuitBreakersResource` and implement `Default` for it too. 
3. Here and in other places use `get_resource_or_default()` to get the resource. This way the resource will always be available, even if there is no circuit breaker layer.
4. Change the `CircuitBreakerChecker` so that it accepts `Arc<CircuitBreakers>` and access circuit breakers through this structure.
5. In the `CircuitBreakerCheckerLayer` you'll get `CircuitBreakersResource` and add a task for circuit breaker checker.

This way the code would work as follows:
- Any task can add a circuit breaker, it doesn't affect anything.
- *If* (and only if) we have a circuit breaker checker layer, we will fetch all the circuit breakers and instantiate a checker task. Otherwise adding these checkers would so nothing, but also (and it's important), would break nothing.

For inspiration, you may look at health check server, which is very similar. We have [AppHealthCheck](https://github.com/matter-labs/zksync-era/blob/main/core/lib/health_check/src/lib.rs#L123) there -- a container to which we can add health checks (for us `CircuitBreakers` would be such a container), and then _separately_ we have a way [to run the healthcheck server](https://github.com/matter-labs/zksync-era/blob/main/core/lib/zksync_core/src/api_server/healthcheck.rs#L52) that accepts `Arc<AppHealthCheck>` (for us it'll be `CircuitBreakerChecker::run(self, breakers: Arc<CircuitBreakers>, stop_receiver: watch::Receiver<bool>)`).

Reply:
Thanks! It fully makes sense.
changed